File: Cpp/189-rotate-array/solution2.cpp
class Solution {
public:
    // TC: O(n)
    // Sc: O(1)
    void rotate(vector<int>& nums, int k) {
        // Get the size of the 'nums' vector
        int n = nums.size();

        // Adjust 'k' to handle cases where 'k' is larger than 'n'
        // This ensures the rotation wraps around if 'k' is larger than 'n'
        k = k % n;

        // Reverse the first part of the vector (0 to n-k-1)
        reverse(nums.begin(), nums.begin() + (n - k));

        // Reverse the second part of the vector (n-k to n-1)
        reverse(nums.begin() + (n - k), nums.end());

        // Reverse the entire vector
        reverse(nums.begin(), nums.end());
    }
};

File: Py/189-rotate-array/solution2.py
from typing import List

class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)

        # Adjust 'k' to handle cases where 'k' is larger than 'n'
        k = k % n

        # Reverse the first part of the list (0 to n-k-1)
        nums[:n-k] = reversed(nums[:n-k])

        # Reverse the second part of the list (n-k to n-1)
        nums[n-k:] = reversed(nums[n-k:])

        # Reverse the entire list
        nums[:] = reversed(nums)
